File: MusicMaker/Core/Abstract/wave.c
#include "wave.h"
#include <math.h>

float play(CWave *wave, float time, float(*shape)(float value)) {
    float position = (M_PI * 2 * wave->frequency) * time;
    return shape(wave->phaseShift + position) * wave->amplitude;
}

float sine_wave(float value) {
    return sinf(value);
}

float sawtooth_wave(float value) {
    return value / (M_PI * 2);
}

float square_wave(float value) {
    return ((int)(truncf(value / M_PI))) % 2 == 0 ? 1 : -1;
}

void normalize_input(float *value) {
    while (*value > M_PI * 2) {
        *value -= M_PI * 2;
    }
}

File: MusicMaker/Scenes/Composer/Abstract/Wave.swift
//
//  Wave.swift
//  MusicMaker
//
//  Created by Sergey Smagleev on 28.05.20.
//  Copyright © 2020 Sergey Smagleev. All rights reserved.
//

import Foundation

class Wave: Tone {
    private let frequency: Float
    private let amplitude: Float
    private let phaseShift: Float

    init(frequency: Float,
         amplitude: Float,
         phaseShift: Float) {
        self.frequency = frequency
        self.amplitude = amplitude
        self.phaseShift = phaseShift
    }
    
    func shape(_ value: Float) -> Float {
        fatalError("this is an abstract class")
    }
    
    func play(time: Float) -> Float {
        let position = (Float.pi * 2 * frequency) * time
        return shape(phaseShift + position) * amplitude
    }
}
